sentry-actix/src/lib.rs

```diff
@@ -299,3 +299,120 @@ fn process_event(
     }
     Some(event)
 }
+
+#[cfg(test)]
+mod tests {
+    use std::io;
+
+    use actix_web::test::{call_service, init_service, TestRequest};
+    use actix_web::{get, web, App, HttpRequest, HttpResponse};
+
+    use sentry::{ClientInitGuard, Level};
+
+    use super::*;
+
+    fn _init() -> ClientInitGuard {
+        // Initialize with a 'valid' DSN so errors are captured
+        sentry::init("https://public@sentry.invalid/1")
+    }
+
+    fn _assert_hub_no_events() {
+        if Hub::current().last_event_id().is_some() {
+            panic!("Current hub should not have had any events.");
+        }
+    }
+
+    fn _assert_hub_has_events() {
+        Hub::current()
+            .last_event_id()
+            .expect("Current hub should have had events.");
+    }
+
+    /// Test explicit events sent to the current Hub inside an Actix service.
+    #[actix_rt::test]
+    async fn test_explicit_events() {
+        let _guard = _init();
+
+        let service = || {
+            // Current hub should have no events
+            _assert_hub_no_events();
+
+            // Add processor to ensure proper event fields
+            sentry::configure_scope(|scope| {
+                scope.add_event_processor(Box::new(move |event| {
+                    let request = event.request.expect("Request should be set.");
+                    assert_eq!(event.transaction, Some("/test".into()));
+                    assert_eq!(event.message, Some("Message".into()));
+                    assert_eq!(event.level, Level::Warning);
+                    assert_eq!(request.method, Some("GET".into()));
+                    None
+                }))
+            });
+
+            // Creates an event
+            sentry::capture_message("Message", Level::Warning);
+
+            // Current Hub should have the event
+            _assert_hub_has_events();
+
+            HttpResponse::Ok()
+        };
+
+        let mut app = init_service(
+            App::new()
+                .wrap(Sentry::new())
+                .service(web::resource("/test").to(service)),
+        )
+        .await;
+
+        // Call the service twice to check the isolation between them
+        for _ in 0..2 {
+            let req = TestRequest::get().uri("/test").to_request();
+            let res = call_service(&mut app, req).await;
```

Comment:
doesn’t the await here make these sequencial instead of concurrent?

Reply:
Yes I was mainly testing the isolation from sequential requests (i.e. making sure the middleware isn't sticky).

Testing the concurrent isolation would be a different beast, I think. How would you determine that both requests used different Hub instances? Is there a way to identify individual Hubs?

Reply:
> Is there a way to identify individual Hubs?

you could use `configure_scope` to set different properties. But I think this is very good now